Multistage Builds

ikaneshiro opened the pull request:
**Description of the Pull Request (PR):**

- Multistage builds
- Some E2E testing for multistage builds
- Some parser changes to accommodate multiple files sections

**This fixes or addresses the following GitHub issues:**

- Fixes #2767 
- Fixes #2768 
- Fixes #2769 

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)


Attn: @singularity-maintainers

ikaneshiro committed:
Changes to Definition struct to allow section args

	This PR changes many of the fields of the Definition struct in
	order to encapsulate the idea of a section having additional
	args. This requires some changes throughout the build package in
	order to accomodate the change. This also changes the parser to
	parse those arguements properly into the Definition struct. Also
	fixes some comments

ikaneshiro committed:
Adding ability for multiple files sections

	This introduces a change to the parser that removes comments
	from lines before parsing for simpler parsing. Also expands
	definition struct to allow multiple files objects to enable
	multistage builds. Unit tests for parser updated to represent
	this struct change.

ikaneshiro committed:
comment fixes

ikaneshiro committed:
Add All() function to parser

	All() will parse all stages in a file and return a slice
	containing definitions of those stages.

ikaneshiro committed:
Add ability to copy files between stages

ikaneshiro committed:
change new test cases to reflect changes in parser package

ikaneshiro committed:
fix gofmt error

ikaneshiro committed:
CLI integration and some edge case checking

ikaneshiro committed:
Merge remote-tracking branch 'upstream/master' into multistage_section_args

ikaneshiro committed:
Some e2e testing for multistage builds

	These tests are mainly to test the copying between stages.

ikaneshiro committed:
fix formatting to make linter happy

cclerget commented on internal/pkg/runtime/engines/imgbuild/create_linux.go:
```diff
@@ -183,8 +185,15 @@ func (engine *EngineOperations) CreateContainer(pid int, rpcConn net.Conn) error
+	files := types.Files{}
+	for _, f := range engine.EngineConfig.Recipe.BuildData.Files {
+		if f.Args == "" {
+			files = f
```
`append` ? Tested it breaks possibility to specify multiple `%files` sections

ikaneshiro replied:
This is actually an error in the parser, will push a fix shortly

cclerget commented on pkg/build/types/parser/deffile.go:
```diff
@@ -244,20 +280,20 @@ func populateDefinition(sections map[string]string, d *types.Definition) (err er
-			Runscript:   sections["runscript"],
-			Test:        sections["test"],
-			Startscript: sections["startscript"],
+			Help:        *sections["help"],
```
Why pass pointer here ?

ikaneshiro replied:
Sections is a map of pointer structs, so this is dereferencing a pointer to a struct so that it can be copied into the final definition struct. I used a pointer in the map so that I could easily modify the struct.

cclerget reviewed (changes requested):
Otherwise looks good !

ikaneshiro committed:
Def parser fix to append multiple %files sections

	Fixes error in parser where consecutive %files sections
	were not appended. A test case has been added to catch this
	bug in the future.

ikaneshiro committed:
removed accidentally added file

WestleyK commented on internal/pkg/build/build.go:
```diff
@@ -337,3 +358,107 @@ func makeDef(spec string, remote bool) (types.Definition, error) {
+
+	d, err := parser.All(defFile)
+	if err != nil {
+		return nil, fmt.Errorf("While parsing definition: %s: %v", spec, err)
```
```suggestion
		return nil, fmt.Errorf("while parsing definition: %s: %v", spec, err)
```

WestleyK commented on internal/pkg/build/build.go:
```diff
@@ -337,3 +358,107 @@ func makeDef(spec string, remote bool) (types.Definition, error) {
+		}
+	}
+
+	return -1, fmt.Errorf("Stage %s was not found", name)
```
Same here:

```suggestion
	return -1, fmt.Errorf("stage %s was not found", name)
```

WestleyK commented on internal/pkg/build/build.go:
```diff
@@ -337,3 +358,107 @@ func makeDef(spec string, remote bool) (types.Definition, error) {
+	copy.Stdout = &output
+	copy.Stderr = &stderr
+	if err := copy.Run(); err != nil {
+		return fmt.Errorf("While copying %v to %v: %v: %v", src, dst, err, stderr.String())
```
```suggestion
		return fmt.Errorf("while copying %v to %v: %v: %v", src, dst, err, stderr.String())
```

WestleyK commented on internal/pkg/build/stage.go:
```diff
@@ -33,41 +32,15 @@ func (s *stage) Assemble(path string) error {
-	if s.b.RunSection("pre") && s.b.Recipe.BuildData.Pre != "" {
+	if s.b.RunSection("pre") && s.b.Recipe.BuildData.Pre.Script != "" {
 		if syscall.Getuid() != 0 {
 			return fmt.Errorf("Attempted to build with scripts as non-root user")
```
```suggestion
 			return fmt.Errorf("attempted to build with scripts as non-root user")
```

WestleyK commented on internal/pkg/runtime/engines/imgbuild/create_linux.go:
```diff
@@ -198,8 +207,10 @@ func (engine *EngineOperations) copyFiles() error {
+		copy.Stderr = &stderr
 		if err := copy.Run(); err != nil {
-			return fmt.Errorf("While copying %v to %v: %v", transfer.Src, transfer.Dst, err)
+			return fmt.Errorf("While copying %v to %v: %v: %v", transfer.Src, transfer.Dst, err, stderr.String())
```
```suggestion
			return fmt.Errorf("while copying %v to %v: %v: %v", transfer.Src, transfer.Dst, err, stderr.String())
```

WestleyK commented on pkg/build/types/parser/deffile.go:
```diff
@@ -131,51 +132,103 @@ func getSectionName(line string) string {
 	if appSections[key] {
 		sectionSplit := strings.SplitN(strings.TrimLeft(split[0], "%"), " ", 3)
 		if len(sectionSplit) < 2 {
 			return fmt.Errorf("App Section %v: Could not be split into section name and app name", sectionSplit[0])
```
```suggestion
 			return fmt.Errorf("app Section %v: could not be split into section name and app name", sectionSplit[0])
```

WestleyK commented on pkg/build/types/parser/deffile.go:
```diff
@@ -131,51 +132,103 @@ func getSectionName(line string) string {
+		// check if first thing parsed is a header/comment or just a section
 		if tok[0] != '%' {
 			if err := doHeader(tok, d); err != nil {
 				return fmt.Errorf("failed to parse DefFile header: %v", err)
```
I dont think error messages are supposed to have any caps...

```suggestion
 				return fmt.Errorf("failed to parse deffile header: %v", err)
```

WestleyK commented on pkg/build/types/parser/deffile.go:
```diff
@@ -350,6 +409,60 @@ func ParseDefinitionFile(r io.Reader) (d types.Definition, err error) {
+
+	raw, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, fmt.Errorf("While attempting to read in definition: %v", err)
```
```suggestion
		return nil, fmt.Errorf("while attempting to read in definition: %v", err)
```

WestleyK reviewed (changes requested):
Left a couple suggestions... Thanks :+1:

ikaneshiro committed:
fix comments

WestleyK commented on internal/pkg/build/build.go:
```diff
@@ -332,8 +353,112 @@ func makeDef(spec string, remote bool) (types.Definition, error) {
+
+	// must be root to build from a definition
+	if os.Getuid() != 0 && !remote {
+		sylog.Fatalf("You must be the root user to build from a Singularity recipe file")
```
Recipe?! Do you mean definition file? :smile: 

This should probably say `recipe`, and to be consistent with are other messages:

```bash
$ singularity build test.sif test.def 
FATAL:   You must be the root user to build from a Singularity recipe file
```

ikaneshiro replied:
@WestleyK are you saying the comment should say recipe or the error should say definition?

WestleyK replied:
Err, misread, got excited...

How about this:

```
FATAL:   You must be the root user to build from a definition file
```

WestleyK commented on internal/pkg/build/build.go:
```diff
@@ -332,8 +353,112 @@ func makeDef(spec string, remote bool) (types.Definition, error) {
+		for _, transfer := range f.Files {
+			// sanity
+			if transfer.Src == "" {
+				sylog.Warningf("Attempt to copy file with no name...")
```
Should this be the same as the other message?

```suggestion
				sylog.Warningf("Attempt to copy file with no name, skipping.")
```

WestleyK reviewed (commented):
Couple more small suggestions...

ikaneshiro committed:
fixed more error messages

WestleyK commented on internal/pkg/build/build.go:
```diff
@@ -332,8 +353,112 @@ func makeDef(spec string, remote bool) (types.Definition, error) {
+
+	// must be root to build from a definition
+	if os.Getuid() != 0 && !remote {
+		sylog.Fatalf("You must be the root user to build definition file")
```
Should this return a error instead of exiting the program? For cleanups or whatever...

WestleyK reviewed (commented):
One question, after that, looks good to me...

ikaneshiro committed:
return error instead of fatal call

WestleyK reviewed (approved):
Looks good to me! :+1:

ikaneshiro committed:
Merge branch 'master' into multistage_stage_copy
